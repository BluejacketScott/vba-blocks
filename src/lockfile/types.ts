import { Snapshot } from '../manifest/types';
import { DependencyGraph } from '../resolve/types';

export interface Lockfile {
  metadata?: { version: string };
  workspace: {
    root: Snapshot;
    members: Snapshot[];
  };
  packages: DependencyGraph;
}
