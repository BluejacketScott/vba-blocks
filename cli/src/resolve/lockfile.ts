import { join } from 'path';
import { exists, readFile, writeFile } from 'fs-extra';
import * as toml from 'toml';
import { Config } from '../config';
import { Version } from '../version';
import { convertToToml } from '../utils';

export interface PackageReference {
  name: string;
  version?: Version;
  source?: string;
  dependencies?: string[];
}

export interface Lockfile {
  root: PackageReference;
  packages: PackageReference[];
  metadata: any;
}

export async function loadLockfile(config: Config): Promise<Lockfile | null> {
  const file = join(config.cwd, 'vba-block.lock');
  if (!await exists(file)) return null;

  const raw = await readFile(file);
  const lockfile = toml.parse(raw);

  return lockfile;
}

export async function writeLockfile(config: Config, lockfile: Lockfile) {
  const file = join(config.cwd, 'vba-block.lock');
  const converted = convertToToml(lockfile);
  await writeFile(file, converted);
}
