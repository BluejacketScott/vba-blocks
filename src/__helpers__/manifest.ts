import { dir as FIXTURES } from '../../tests/__fixtures__';
import { Manifest } from '../manifest';
import { isPathDependency } from '../manifest/dependency';
import { relative } from '../utils/path';

export function normalizeManifest(manifest: Manifest, relativeTo: string = FIXTURES): Manifest {
  for (const src of manifest.src) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.dependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }
  for (const src of manifest.devSrc) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.devDependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }

  if (manifest.target) {
    manifest.target.path = normalizePath(manifest.target.path, relativeTo);
  }

  return manifest;
}

function normalizePath(path: string, relativeTo: string = FIXTURES): string {
  return relative(relativeTo, path);
}
