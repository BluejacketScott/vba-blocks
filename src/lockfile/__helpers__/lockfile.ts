export function normalizeLockfile(lockfile: string): string {
	return lockfile.replace(/vba\-blocks v\d.\d.\d/g, "vba-blocks v#.#.#");
}
