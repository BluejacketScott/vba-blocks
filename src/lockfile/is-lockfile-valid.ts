import { satisfies } from 'semver';
import { loadManifest, Manifest } from '../manifest';
import {
  Dependency,
  isGitDependency,
  isPathDependency,
  isRegistryDependency
} from '../manifest/dependency';
import { Workspace } from '../professional/workspace';
import has from '../utils/has';
import { Lockfile, LOCKFILE_VERSION, MinimalSnapshot } from './lockfile';

/**
 * ## Is lockfile valid?
 *
 * - [ ] Lockfile version has not changed
 * - [ ] Workspace root manifest has not changed
 * - [ ] Workspace members have not changed
 */
export default async function isLockfileValid(
  lockfile: Lockfile,
  workspace: Workspace
): Promise<boolean> {
  if (!lockfile.metadata || lockfile.metadata.version !== LOCKFILE_VERSION) return false;

  if (!(await hasManifestChanged(workspace.root, lockfile.workspace.root))) return false;

  if (lockfile.workspace.members.length !== workspace.members.length) return false;

  const membersByName: { [name: string]: Manifest } = {};
  workspace.members.forEach(member => (membersByName[member.name] = member));

  for (const lockedMember of lockfile.workspace.members) {
    const currentMember = membersByName[lockedMember.name];
    if (!currentMember) return false;
    if (!(await hasManifestChanged(currentMember, lockedMember))) return false;
  }

  return true;
}

/**
 * ## Has manifest changed?
 *
 * - [ ] Has name changed?
 * - [ ] Have dependencies been added or removed?
 * - [ ] Are dependencies satisfied by lockfile versions?
 */
async function hasManifestChanged(current: Manifest, locked: MinimalSnapshot): Promise<boolean> {
  if (current.name !== locked.name) return false;

  return await haveDependenciesChanged(current, locked);
}

async function haveDependenciesChanged(
  current: Manifest,
  locked: MinimalSnapshot
): Promise<boolean> {
  const dependencies = current.dependencies.concat(current.devDependencies);

  if (dependencies.length !== locked.dependencies.length) return false;

  const dependenciesByName: { [name: string]: Dependency } = {};
  dependencies.forEach(dependency => {
    dependenciesByName[dependency.name] = dependency;
  });

  for (const lockedDependency of locked.dependencies) {
    const currentDependency = dependenciesByName[lockedDependency.name];
    if (!currentDependency) return false;
    if (!(await isDependencySatisfied(currentDependency, lockedDependency))) return false;
  }

  return true;
}

async function isDependencySatisfied(current: Dependency, locked: Dependency): Promise<boolean> {
  if (isRegistryDependency(locked)) {
    // Note: lockfile value is more specific
    // -> compare locked to current manifest's value
    return isRegistryDependency(current) && satisfies(locked.version, current.version);
  } else if (isPathDependency(locked)) {
    if (!isPathDependency(current)) return false;
    if (current.path !== locked.path) return false;

    // Load current version from manifest of path dependency
    const manifest = await loadManifest(current.path);
    return manifest.version === locked.version!;
  } else if (isGitDependency(locked)) {
    if (!isGitDependency(current)) return false;

    if (has(current, 'rev') && has(locked, 'rev')) return current.rev === locked.rev;
    if (has(current, 'tag') && has(locked, 'tag')) return current.tag === locked.tag;
    if (has(current, 'branch') && has(locked, 'branch')) return current.branch === locked.branch;
  }

  return false;
}
