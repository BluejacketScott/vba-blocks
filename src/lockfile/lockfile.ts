import { Dependency } from "../manifest/dependency";
import { DependencyGraph } from "../resolve/dependency-graph";

export interface MinimalSnapshot {
	name: string;
	dependencies: Dependency[];
}

export const LOCKFILE_VERSION = "1";

/**
 * ## Lockfile
 *
 * Save resolved dependency graph and workspace information
 * for precisely rebuilding dependency graph if the workspace hasn't changed
 *
 * Version the lockfile separately from vba-blocks to allow the lockfile format to change over time,
 * while staying isolated from the more-frequent changes to the vba-blocks version.
 */
export interface Lockfile {
	metadata?: { version: string };
	workspace: {
		root: MinimalSnapshot;
		members: MinimalSnapshot[];
	};
	packages: DependencyGraph;
}
