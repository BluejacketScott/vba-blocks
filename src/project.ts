import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { Config, loadConfig } from './config';
import { Manifest, loadManifest } from './manifest';
import { Workspace, loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve, { DependencyGraph } from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';

const debug = require('debug')('vba-blocks:project');

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
    staging: string;
  };
  has_dirty_lockfile: boolean;
}

/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir: string = env.cwd): Promise<Project> {
  const manifest = await loadManifest(dir);

  const config = await loadConfig();
  const workspace = await loadWorkspace(manifest);
  const lockfile = await readLockfile(workspace.root.dir);

  // Resolve packages from lockfile or from sources
  const has_dirty_lockfile = !lockfile || !isLockfileValid(lockfile, workspace);
  debug(
    !has_dirty_lockfile
      ? 'Loading packages from lockfile'
      : 'Resolving packages'
  );

  const packages = !has_dirty_lockfile
    ? lockfile!.packages
    : await resolve(config, workspace, lockfile ? lockfile.packages : []);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup'),
    staging: await tmpFolder({ dir: env.staging })
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths,
    has_dirty_lockfile
  };
}

export interface FetchProject {
  manifest: Manifest;
  packages: DependencyGraph;
  config: Config;
}

/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(
  project: FetchProject
): Promise<Manifest[]> {
  const manifests = await parallel(
    project.packages,
    async registration => {
      const path = await fetch(project.config.sources, registration);
      const manifest = await loadManifest(path);

      return manifest;
    },
    { progress: env.reporter.progress('Fetch dependencies') }
  );

  return manifests;
}

export interface ProjectOptions {
  type?: 'project' | 'package';
}

/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(
  name: string,
  dir: string,
  options: ProjectOptions = {}
): Promise<Project> {
  dir = normalize(dir);
  const { type = 'project' } = options;

  const config = await loadConfig();

  // TODO load defaults from config
  const version = '0.1.0';
  const authors: string[] = [];
  const license = 'MIT';

  // Manually generate manifest and project
  // (may be included in project in the future)
  const manifest: Manifest = {
    type,
    name,
    version,
    metadata: {
      authors,
      publish: type === 'package',
      license
    },
    dependencies: [],
    src: [],
    references: [],
    targets: [],
    dir
  };

  const workspace = await loadWorkspace(manifest);

  const project: Project = {
    manifest,
    workspace,
    packages: [],
    config,
    paths: {
      root: workspace.root.dir,
      dir,
      build: join(dir, 'build'),
      backup: join(dir, 'build', '.backup'),
      staging: await tmpFolder({ dir: env.staging })
    },
    has_dirty_lockfile: true
  };

  return project;
}
