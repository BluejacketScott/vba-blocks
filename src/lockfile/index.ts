import { ok } from 'assert';
import env from '../env';
import { CliError, ErrorCode } from '../errors';
import { Manifest, Snapshot } from '../manifest';
import { Dependency } from '../manifest/dependency';
import { DEFAULT_VERSION } from '../manifest/version';
import { Project } from '../project.js';
import { getRegistration } from '../resolve';
import { DependencyGraph } from '../resolve/dependency-graph';
import {
  getRegistrationId,
  getRegistrationSource,
  getSourceParts,
  Registration,
  toDependency
} from '../sources/registration';
import { pathExists, readFile, writeFile } from '../utils/fs';
import has from '../utils/has';
import { join, relative, trailing } from '../utils/path';
import { parse as parseToml, toLockfile as convertToLockfileToml } from '../utils/toml';
import { Lockfile, LOCKFILE_VERSION } from './lockfile';

const debug = env.debug('vba-blocks:lockfile');

type DependencyByName = Map<string, Dependency>;

export { default as isLockfileValid } from './is-lockfile-valid';

/**
 * Read lockfile at given dir (if present)
 * (for invalid lockfile, errors are ignored and treated as no lockfile)
 */
export async function readLockfile(dir: string): Promise<Lockfile | null> {
  const file = join(dir, 'vba-block.lock');
  debug(`Reading lockfile from ${file}`);

  if (!(await pathExists(file))) {
    debug(`Not found`);
    return null;
  }

  try {
    const toml = await readFile(file, 'utf8');
    return await fromToml(toml, dir);
  } catch (err) {
    debug(`Error reading/parsing`);
    debug(err);

    return null;
  }
}

/**
 * Write lockfile for project to given dir
 *
 * @throws LockfileWriteFailed
 */
export async function writeLockfile(dir: string, project: Project): Promise<void> {
  const file = join(dir, 'vba-block.lock');
  debug(`Writing lockfile to ${file}`);

  try {
    const toml = toToml(project, dir);
    await writeFile(file, toml);
  } catch (err) {
    debug('Error converting/writing');
    debug(err);

    throw new CliError(ErrorCode.LockfileWriteFailed, `Failed to write lockfile to "${file}".`);
  }
}

/**
 * Convert project to lockfile toml, alphabetized and with trailing commas
 * (should be suitable for VCS)
 */
export function toToml(project: Project, dir: string): string {
  const root = toLockfileManifest(project.workspace.root, project.packages, dir);
  const members: any[] = project.workspace.members.map(member =>
    toLockfileManifest(member, project.packages, dir)
  );

  const packages: any[] = project.packages.sort(byPackageName).map((registration: Registration) => {
    const { name, version, source } = registration;
    const dependencies = registration.dependencies.map(dependency =>
      toDependencyId(dependency, project.packages, dir)
    );

    return {
      name,
      version,
      source: toLockfileSource(source, dir),
      dependencies
    };
  });

  const metadata = { version: LOCKFILE_VERSION };
  const toml = convertToLockfileToml({ metadata, root, members, packages });

  return `${toml}`;
}

/**
 * Load lockfile from toml (including "hydrating" dependencies from packages)
 */
export async function fromToml(toml: string, dir: string): Promise<Lockfile> {
  const parsed = await parseToml(toml);
  ok(has(parsed, 'root'), 'vba-block.lock is missing [root] field');

  const metadata = parsed.metadata;

  // First pass through packages to load high-level information
  // (needed to map dependencies to parsed packages)
  const byName: DependencyByName = new Map();
  const packages = (parsed.packages || []).map((value: any) => {
    const { name, version, source, dependencies } = value;
    ok(name && version && source && Array.isArray(dependencies), 'Invalid package in lockfile');

    const registration = {
      id: getRegistrationId(name, version),
      name,
      version,
      source: toSource(source, dir),
      dependencies
    };

    byName.set(name, toDependency(registration));

    return registration;
  });

  // Hydrate dependencies of packages
  packages.forEach((registration: any) => {
    registration.dependencies = registration.dependencies.map((id: string) =>
      getDependency(id, byName)
    );
  });

  // Load manifests for workspace
  const root = toManifest(parsed.root, byName);
  const members: Snapshot[] = (parsed.members || []).map((member: any) =>
    toManifest(member, byName)
  );

  return { metadata, workspace: { root, members }, packages };
}

/**
 * Convert raw toml value to manifest
 */
function toManifest(value: any, dependencyByName: DependencyByName): Snapshot {
  const { name, version } = value;
  ok(name && version && Array.isArray(value.dependencies), 'Invalid manifest in lockfile');

  // "Hydrate" dependencies from ids
  const dependencies = value.dependencies.map((id: string) => getDependency(id, dependencyByName));

  return {
    name,
    version,
    dependencies
  };
}

interface LockfileManifest {
  name: string;
  version?: string;
  dependencies: string[];
}

/**
 * Get toml from manifest
 */
function toLockfileManifest(
  manifest: Manifest,
  packages: DependencyGraph,
  dir: string
): LockfileManifest {
  const { name, version } = manifest;

  // Convert dependencies and dev-dependencies to ids
  const all_dependencies = manifest.dependencies.concat(manifest.devDependencies);
  const dependencies = all_dependencies.map(dependency =>
    toDependencyId(dependency, packages, dir)
  );

  if (version === DEFAULT_VERSION) {
    return { name, dependencies };
  }

  return {
    name,
    version,
    dependencies
  };
}

/**
 * Get toml from source string
 * (Convert full path to relative to root dir)
 */
function toLockfileSource(source: string, dir: string): string {
  const { type, value, details } = getSourceParts(source);
  if (type !== 'path') return source;

  const relativePath = trailing(relative(dir, value));
  return getRegistrationSource(type, relativePath, details);
}

/**
 * "Hydrate" registration source from raw toml
 * (Convert relative paths to absolute)
 */
function toSource(source: string, dir: string): string {
  const { type, value, details } = getSourceParts(source);
  if (type !== 'path') return source;

  const absolutePath = join(dir, value);
  return getRegistrationSource(type, absolutePath, details);
}

/**
 * Convert dependency to unique id
 *
 * Minimum information needed for lockfile:
 * "{name} {version} {source}"
 */
function toDependencyId(dependency: Dependency, packages: DependencyGraph, dir: string) {
  const registration = getRegistration(packages, dependency);
  ok(registration, `No package found for dependency "${dependency.name}"`);

  let { version, source } = registration!;
  const { type, value, details } = getSourceParts(source);

  if (type === 'path') {
    const relativePath = trailing(relative(dir, value));
    source = getRegistrationSource(type, relativePath, details);
  }

  return `${dependency.name} ${version} ${source}`;
}

function getDependency(id: string, dependencyByName: DependencyByName): Dependency {
  const [name] = id.split(' ', 1);
  const dependency = dependencyByName.get(name);

  ok(dependency, `Package not found in lockfile, "${id}"`);

  return dependency!;
}

function byPackageName(a: Registration, b: Registration) {
  if (a.name < b.name) return -1;
  if (a.name > b.name) return 1;
  return 0;
}
