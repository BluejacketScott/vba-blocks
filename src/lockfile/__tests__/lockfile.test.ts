import { toToml } from '../';
import { dev } from '../../../tests/__fixtures__';
import { setup } from '../../../tests/__helpers__/project';
import { normalizeLockfile } from '../__helpers__/lockfile';

test('loads and parses manifest', async () => {
  const { project } = await setup(dev);

  const toml = toToml(project, dev);

  expect(normalizeLockfile(toml)).toMatchSnapshot();
});
