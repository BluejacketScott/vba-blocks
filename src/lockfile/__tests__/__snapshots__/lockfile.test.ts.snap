// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and parses manifest 1`] = `
"# Auto-generated by vba-blocks v#.#.#
[metadata]
version = \\"1\\"

[root]
name = \\"dev\\"
dependencies = [
  \\"dictionary 1.4.1 registry+vba-blocks#sha256-b241062bcd7ffd68c9063e3980a5df4c6781df29d96763a020d291e52454f6a5\\",
  \\"json 2.3.0 registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799\\",
]

[[packages]]
name = \\"dictionary\\"
version = \\"1.4.1\\"
source = \\"registry+vba-blocks#sha256-b241062bcd7ffd68c9063e3980a5df4c6781df29d96763a020d291e52454f6a5\\"
dependencies = [
]

[[packages]]
name = \\"json\\"
version = \\"2.3.0\\"
source = \\"registry+vba-blocks#sha256-3c2760ce2940dd19c16542cead792e0d05e58566155b1fbc72d445fe50a36799\\"
dependencies = [
  \\"dictionary 1.4.1 registry+vba-blocks#sha256-b241062bcd7ffd68c9063e3980a5df4c6781df29d96763a020d291e52454f6a5\\",
]

"
`;
